atomic_reactor/plugins/pre_fetch_sources.py

```diff
@@ -0,0 +1,150 @@
+of the BSD license. See the LICENSE file for details.
+"""
+
+from __future__ import absolute_import
+
+import os
+
+from six.moves.urllib.parse import urlparse
+
+from atomic_reactor.constants import PLUGIN_FETCH_SOURCES_KEY, DEFAULT_DOWNLOAD_BLOCK_SIZE
+from atomic_reactor.plugin import PreBuildPlugin
+from atomic_reactor.plugins.pre_reactor_config import (
+    get_koji,
+    get_koji_path_info,
+    get_koji_session,
+    NO_FALLBACK
+)
+from atomic_reactor.util import get_retrying_requests_session
+
+
+class FetchSourcesPlugin(PreBuildPlugin):
+    key = PLUGIN_FETCH_SOURCES_KEY
+    is_allowed_to_fail = False
+    DOWNLOAD_DIR = 'image_sources'
+
+    def __init__(
+        self, tasker, workflow, koji_build_id=None, koji_build_nvr=None, signing_intent=None,
+        kojihub=None, kojiroot=None, koji_proxy_user=None, koji_ssl_certs=None, koji_principal=None,
+        koji_keytab=None
+    ):
+        """
+        :param tasker: ContainerTasker instance
+        :param workflow: DockerBuildWorkflow instance
+        :param build_id: int, container image koji build id
+        :param build_nvr: str, container image koji build NVR
+        :param signing_intent: str, TODO
+        :param koji_build_id: int, container image koji build id
+        :param kojihub: str, koji hub (xmlrpc)
+        :param kojiroot: str, Koji root storage URL
+        :param koji_proxy_user: str, user to log in as (requires hub config)
+        :param koji_ssl_certs_dir: str, path to 'cert', 'ca', 'serverca'
+        :param koji_principal: str, Kerberos principal (must specify keytab)
+        :param koji_keytab: str, keytab name (must specify principal)
+        """
+        # TODO: should we add a 'sources' build type and skip this plugin if not a sources build?
+        if not koji_build_id and not koji_build_nvr:
+            err_msg = ('{} expects either koji_build_id or koji_build_nvr to be defined'
+                       .format(self.__class__.__name__))
+            raise TypeError(err_msg)
+        if koji_build_id is not None and not isinstance(koji_build_id, int):
+            err_msg = 'koji_build_id must be an int. Got {}'.format(type(koji_build_id))
+            raise TypeError(err_msg)
+        if koji_build_nvr is not None and not isinstance(koji_build_nvr, str):
+            err_msg = 'koji_build_id must be an str. Got {}'.format(type(koji_build_nvr))
+            raise TypeError(err_msg)
+
+        super(FetchSourcesPlugin, self).__init__(tasker, workflow)
+        self.koji_build_id = koji_build_id
+        self.koji_build_nvr = koji_build_nvr
+        self.signing_intent = signing_intent
+        self.koji_fallback = {
+            'hub_url': kojihub,
+            'root_url': kojiroot,
+            'auth': {
+                'proxyuser': koji_proxy_user,
+                'ssl_certs_dir': koji_ssl_certs,
+                'krb_principal': str(koji_principal),
+                'krb_keytab_path': str(koji_keytab)
```

Comment:
Should `koji_principal` and `koji_keytab` default to the empty string instead of `None`?